File: server/src/chat/routes/scenarios.ts
/**
 * scenarios.ts — REST endpoints for scenario CRUD
 *
 * Mirrors the Python backend's /scenarios endpoints.
 * Scenarios live in `scenarios/{id}.json`.
 *
 * Prefix: /chat/api  (applied externally when mounting)
 */

import { Router } from "express";
import * as fs from "node:fs/promises";
import * as fsSync from "node:fs";
import * as path from "node:path";

// ── Router ──────────────────────────────────────────────────────────────────

export function createScenarioRoutes(projectRoot: string): Router {
  const router = Router();
  const scenariosDir = path.join(projectRoot, "scenarios");

  // ── GET /scenarios ────────────────────────────────────────────────────────
  router.get("/scenarios", async (_req, res) => {
    try {
      let entries: string[];
      try {
        entries = await fs.readdir(scenariosDir);
      } catch {
        res.json([]);
        return;
      }

      entries.sort();
      const result: Array<Record<string, unknown>> = [];

      for (const entry of entries) {
        if (!entry.endsWith(".json")) continue;
        const filePath = path.join(scenariosDir, entry);
        try {
          const raw = await fs.readFile(filePath, "utf-8");
          result.push(JSON.parse(raw) as Record<string, unknown>);
        } catch {
          // skip malformed
        }
      }

      res.json(result);
    } catch (err: any) {
      res.status(500).json({ error: err.message ?? "Internal error" });
    }
  });

  // ── GET /scenarios/:id ────────────────────────────────────────────────────
  router.get("/scenarios/:id", async (req, res) => {
    try {
      const scenarioId = req.params.id;
      const filePath = path.join(scenariosDir, `${scenarioId}.json`);

      try {
        const raw = await fs.readFile(filePath, "utf-8");
        res.json(JSON.parse(raw) as Record<string, unknown>);
      } catch {
        res.status(404).json({ error: "Scenario not found" });
      }
    } catch (err: any) {
      res.status(500).json({ error: err.message ?? "Internal error" });
    }
  });

  // ── POST /scenarios ──────────────────────────────────────────────────────
  router.post("/scenarios", async (req, res) => {
    try {
      const body = req.body as Record<string, unknown>;
      const sid = ((body.id as string) ?? "").trim();
      if (!sid || /[/\\.\s]/.test(sid) || sid.length > 64) {
        res.status(400).json({ error: "Invalid scenario id" });
        return;
      }

      await fs.mkdir(scenariosDir, { recursive: true });
      const filePath = path.join(scenariosDir, `${sid}.json`);

      if (fsSync.existsSync(filePath)) {
        res
          .status(409)
          .json({ error: `Scenario '${sid}' already exists` });
        return;
      }

      const data = {
        id: sid,
        name: (body.name as string) ?? sid,
        description: (body.description as string) ?? "",
        system_prompt: (body.system_prompt as string) ?? "",
        suggested_agents: (body.suggested_agents as string[]) ?? [],
        topic_hint: (body.topic_hint as string) ?? "",
      };

      await fs.writeFile(
        filePath,
        JSON.stringify(data, null, 2),
        "utf-8",
      );
      res.json({ ok: true, id: sid });
    } catch (err: any) {
      res.status(500).json({ error: err.message ?? "Internal error" });
    }
  });

  // ── PUT /scenarios/:id ───────────────────────────────────────────────────
  router.put("/scenarios/:id", async (req, res) => {
    try {
      const scenarioId = req.params.id;
      const filePath = path.join(scenariosDir, `${scenarioId}.json`);

      let data: Record<string, unknown>;
      try {
        const raw = await fs.readFile(filePath, "utf-8");
        data = JSON.parse(raw) as Record<string, unknown>;
      } catch {
        res.status(404).json({ error: "Scenario not found" });
        return;
      }

      const body = req.body as Record<string, unknown>;
      for (const key of [
        "name",
        "description",
        "system_prompt",
        "suggested_agents",
        "topic_hint",
      ]) {
        if (key in body) data[key] = body[key];
      }

      await fs.writeFile(
        filePath,
        JSON.stringify(data, null, 2),
        "utf-8",
      );
      res.json({ ok: true });
    } catch (err: any) {
      res.status(500).json({ error: err.message ?? "Internal error" });
    }
  });

  // ── DELETE /scenarios/:id ────────────────────────────────────────────────
  router.delete("/scenarios/:id", async (req, res) => {
    try {
      const scenarioId = req.params.id;
      const filePath = path.join(scenariosDir, `${scenarioId}.json`);

      try {
        await fs.access(filePath);
      } catch {
        res.status(404).json({ error: "Scenario not found" });
        return;
      }

      await fs.unlink(filePath);
      res.json({ ok: true });
    } catch (err: any) {
      res.status(500).json({ error: err.message ?? "Internal error" });
    }
  });

  return router;
}

File: routes/scenarios.py
"""Scenario CRUD endpoints."""
import json
import re

from fastapi import APIRouter, HTTPException

router = APIRouter()


@router.get("/scenarios")
async def list_scenarios():
    """Return all scenario JSON files from the scenarios/ directory."""
    import app as _app
    if not _app.SCENARIOS_DIR.exists():
        return []
    result = []
    for f in sorted(_app.SCENARIOS_DIR.glob("*.json")):
        try:
            result.append(json.loads(f.read_text()))
        except Exception:
            pass
    return result


@router.get("/scenarios/{scenario_id}")
async def get_scenario(scenario_id: str):
    import app as _app
    f = _app.SCENARIOS_DIR / f"{scenario_id}.json"
    if not f.exists():
        raise HTTPException(status_code=404, detail="Scenario not found")
    return json.loads(f.read_text())


@router.post("/scenarios")
async def create_scenario(body: dict = {}):
    import app as _app
    sid = (body.get("id") or "").strip()
    if not sid or re.search(r'[/\\.\s]', sid) or len(sid) > 64:
        raise HTTPException(status_code=400, detail="Invalid scenario id")
    _app.SCENARIOS_DIR.mkdir(parents=True, exist_ok=True)
    f = _app.SCENARIOS_DIR / f"{sid}.json"
    if f.exists():
        raise HTTPException(status_code=409, detail=f"Scenario '{sid}' already exists")
    data = {
        "id": sid,
        "name": body.get("name", sid),
        "description": body.get("description", ""),
        "system_prompt": body.get("system_prompt", ""),
        "suggested_agents": body.get("suggested_agents", []),
        "topic_hint": body.get("topic_hint", ""),
    }
    f.write_text(json.dumps(data, indent=2, ensure_ascii=False))
    return {"ok": True, "id": sid}


@router.put("/scenarios/{scenario_id}")
async def update_scenario(scenario_id: str, body: dict = {}):
    import app as _app
    f = _app.SCENARIOS_DIR / f"{scenario_id}.json"
    if not f.exists():
        raise HTTPException(status_code=404, detail="Scenario not found")
    data = json.loads(f.read_text())
    for key in ("name", "description", "system_prompt", "suggested_agents", "topic_hint"):
        if key in body:
            data[key] = body[key]
    f.write_text(json.dumps(data, indent=2, ensure_ascii=False))
    return {"ok": True}


@router.delete("/scenarios/{scenario_id}")
async def delete_scenario(scenario_id: str):
    import app as _app
    f = _app.SCENARIOS_DIR / f"{scenario_id}.json"
    if not f.exists():
        raise HTTPException(status_code=404, detail="Scenario not found")
    f.unlink()
    return {"ok": True}
